rust/datafusion/src/physical_plan/datetime_expressions.rs

```diff
@@ -205,6 +206,108 @@ pub fn to_timestamp(args: &[ArrayRef]) -> Result<TimestampNanosecondArray> {
     Ok(TimestampNanosecondArray::from(Arc::new(data)))
 }
 
+/// date_trunc SQL function
+pub fn date_trunc(args: &[ArrayRef]) -> Result<TimestampNanosecondArray> {
+    let granularity_array =
+        &args[0]
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .ok_or_else(|| {
+                DataFusionError::Execution(
+                    "Could not cast date_trunc granularity input to StringArray"
+                        .to_string(),
+                )
+            })?;
+
+    let array = &args[1]
+        .as_any()
+        .downcast_ref::<TimestampNanosecondArray>()
+        .ok_or_else(|| {
+            DataFusionError::Execution(
+                "Could not cast date_trunc array input to TimestampNanosecondArray"
+                    .to_string(),
+            )
+        })?;
+
+    let range = 0..array.len();
+    let result = range
+        .map(|i| {
+            if array.is_null(i) {
+                Ok(0_i64)
+            } else {
+                let date_time = match granularity_array.value(i) {
```

Comment:
This list seems consistent with https://www.postgresql.org/docs/9.1/functions-datetime.html#FUNCTIONS-DATETIME-TRUNC 👍 I think it is fine that we don't yeet support some of the more esoteric stuff (like milleniumm')